### ISSUE TITLE
"Project is corrupt (Unable to work with the blockfiles)" when saving to Mac directory saved on iCloud

### ISSUE THREAD

**Issue by peterjdann** [audacity/audacity#5577]
### Bug description

I am able to save, and then reopen, a 25 minute long, single track Audacity 3.4.1 project on a local folder that is NOT associated with iCloud Drive. However, if I try to save, and then reopen, a similar 25 minute long Audacity 3.4.1 project on a folder that IS associated with iCloud drive, I get the message described in the title.

I have noticed that this problem does not occur if I record only a 30 second project.

Detailed report is:

{
    "timestamp": 1699672480,
    "event_id": "28bd7f49faf443ea856b78c0d486418f",
    "platform": "native",
    "release": "audacity@3.4.1",
    "contexts": {
        "os": {
            "type": "os",
            "name": "Macintosh",
            "version": "10.16.0"
        }
    },
    "exception": {
        "values": [
            {
                "type": "Error_Opening_Project",
                "value": "Project is corrupt\n(Unable to work with the blockfiles)",
                "mechanism": {
                    "type": "runtime_error",
                    "handled": false,
                    "data": {
                        "sqlite3.query": "DELETE FROM sampleblocks WHERE NOT inset(blockid);",
                        "sqlite3.rc": "11",
                        "sqlite3.context": "ProjectGileIO::GetBlob",
                        "log": "Error: Project is corrupt\n(Unable to work with the blockfiles)\n14:14:40: Loaded 67 string <path> Kb in size\n14:14:40: sqlite3 message: (11) database corruption at line 67162 of [1b256d97b5]\n14:14:40: sqlite3 message: (11) statement aborts at 9: [DELETE FROM sampleblocks WHERE NOT inset(blockid);] database disk image is malformed\n14:14:40: DBConnection SetDBError\n\tErrorCode: 11\n\tLastError: Project is corrupt\n(Unable to work with the blockfiles)\n\tLibraryError: database disk image is malformed\n"
                    }
                }
            }
        ]
    }
}

Until shortly before performing these tests, I had my iCloud Mac Storage set to "optimised". Just before performing these tests, I turned that setting off.
<img width="385" alt="Error mes
<img width="1103" alt="Open recent files list" src="https://github.com/audacity/audacity/assets/65576684/5d9b1687-3404-494e-a3db-dbc26f7c846a">
sage" src="https://github.com/audacity/audacity/assets/65576684/0f581ef3-9e5e-4f23-b0ad-85221aaa02c5">

### Steps to reproduce

1. On a Mac running Sonoma and with iCloud drive active, record a 25 minute single track project in Audacity 3.4.1 and save it to a subfolder of your Documents directory. (This should cause the project to be saved on iCloud drive.)
2. Attempt to open the project.

### Expected behavior

I should be able to open the project, even though it is stored on an iCloud Drive. Ten days or so ago, I was able to do so (for example, when recording the files that include "The Charterhouse of Parma" in their filenames visible in my screenshot).

### Actual behavior

An error message indicating the project file is corrupt (see image attached above).

### Audacity Version

latest stable version (from audacityteam.org/download)

### Operating system

macOS

### Additional context

Audacity version: 3.4.1

OS: Sonoma 14.0

Machine: iMac Retina 5K, 27 inch, 2019. Memory 16 GB.

Recording device is a Focusrite sound module to which I'm connected by a Thunderbolt/USB cable.

**Comment by peterjdann** [comment]
I see my screenshot of error message did not appear. I reattach it her.
<img width="385" alt="Error message" src="https://github.com/audacity/audacity/assets/65576684/eb04d498-bb0c-401c-8794-7b93ab716343">

**Comment by JohnColket** [comment]
The underlying database handler in Audacity, SQLite does not work perfectly with networked drives.  For a discussion on this, read: https://www.sqlite.org/useovernet.html. As long as you are using networked drives you are going to have these issues.

If you have a file that locally exhibits, this "sqlite3.rc": "11", error code 11=Corrupt Database file, there is a good chance you can rescue it using the procedure outlined here: https://forum.audacityteam.org/t/corrupt-or-otherwise-broken-audacity-project-recovery/64162

**Comment by peterjdann** [comment]
Thanks John. 

That utility, as you will know, does not work on a Mac. I was able to use linux version to retrieve a fair portion (but not all) of a corrupted project.

It appears to me that Apple's handling of what you are calling (in iCloud) a "networked drive" has changed with Sonoma — everything worked fine before Sonoma, even with iCloud drive, for me — but that's not really the issue here.
Assuming for the moment that there's no way that SQLite can be got to work reliably with a Mac folder managed by iCloud under Sonoma, Audacity needs to handle what is a potentially dangerous situation better than it does. Right now, Audacity (the software, I mean) gives the user the impression their data is being saved safely in a scenario where it's actually not, resulting in significant data loss that there is not even a Mac-compatible utility to mitigate.

I'm not a programmer, but I have to ask the question: is there some way of determining in the software at the time a user is proposing to save a project to a certain location whether SQLite is going to able to perform properly with respect to that destination?  In my own case (see attached screenshots above) I would guess that the presence in the path of the string "com-apple-CloudDocs" might be a pretty strong indication that the file saving destination is going to be unsuitable.

A safe and reliable sound recording and editing utility shouldn't have to rely on users knowing abstruse details about how its inner working behave in certain conditions, but instead should protect them from doing what is plainly dangerous in the first place.

**Comment by peterjdann** [comment]
Today I have installed Sonoma version 14.1.1 on my Mac, and this problem appears to have gone away. (I experienced this problem when running Sonoma 14.0.) Interestingly, the path to any subfolder of my Documents folder no longer shows up in the Aucacity recent files list as including "com-apple-CloudDocs" (with tilde characters, not hyphens), but has reverted to its pre-Sonoma 14.0 appearance. It looks to me like Apple somewhat screwed up their implementation of iCloud with Sonoma 14.0, but have fixed at least some of these problems now.
